File: C++/代码随想录/Binary_Tree/Find the attributes of the binary search tree/700. Search in a Binary Search Tree.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def searchBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if root is None or root.val == val:
            return root
        elif root.val < val:
            return self.searchBST(root.right, val)
        else:
            return self.searchBST(root.left, val)

# Test case
root = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7))
sol = Solution()
node = sol.searchBST(root, 2)
print(node.val if node else "Not Found")  # 2

File: C++/代码随想录/Binary_Tree/Find the attributes of the binary search tree/700. Search in a Binary Search Tree.cpp
// 700. Search in a Binary Search Tree
// https://leetcode.com/problems/search-in-a-binary-search-tree/
// clang++ 700.\ Search\ in\ a\ Binary\ Search\ Tree.cpp -std=c++17
#include "../../printUtils.hpp"

template <typename T>
TreeNode<T>* searchBST(TreeNode<T>* root, T val) {
    if (!root) return nullptr;
    if (root->val == val) return root;
    if (root->val > val) return searchBST(root->left, val);
    return searchBST(root->right, val);
}

int main() {
    TreeNode<int>* root = buildTree({4, 2, 7, 1, 3}, -1);
    int target = 2;
    TreeNode<int>* node = searchBST(root, target);

    std::cout << "The subtree rooted with the node with value " << target << " is: " << std::endl;
    printArray(levelOrder(node));

    return 0;
}